### ISSUE TITLE
Question regarding choice of reward function for AIRL training

### ISSUE THREAD

**Issue by Peter-Nguyen-Duc** [ku2482/gail-airl-ppo.pytorch#14]
hello, first off, your code repo is super awesome, works like a charm.

I actually just want to ask about the choice of reward function when training the RL agent within the AIRL loop.

It seems as the RL agent gets a reward function calculated as: "-logsigmoid(-logits)" which is "-log(1 - D)." where D is a discriminator from [0,1].
The code is within the function "calculate_reward" underneath the class AIRLDiscrim as shown below:

path: gail_airl_ppo/network/disc.py
`
class AIRLDiscrim(nn.Module):

    def __init__(self, state_shape, gamma,
                 hidden_units_r=(64, 64),
                 hidden_units_v=(64, 64),
                 hidden_activation_r=nn.ReLU(inplace=True),
                 hidden_activation_v=nn.ReLU(inplace=True)):
        super().__init__()

        self.g = build_mlp(
            input_dim=state_shape[0],
            output_dim=1,
            hidden_units=hidden_units_r,
            hidden_activation=hidden_activation_r
        )
        self.h = build_mlp(
            input_dim=state_shape[0],
            output_dim=1,
            hidden_units=hidden_units_v,
            hidden_activation=hidden_activation_v
        )

        self.gamma = gamma

    def f(self, states, dones, next_states):
        rs = self.g(states)
        vs = self.h(states)
        next_vs = self.h(next_states)
        return rs + self.gamma * (1 - dones) * next_vs - vs

    def forward(self, states, dones, log_pis, next_states):
        # Discriminator's output is sigmoid(f - log_pi).
        return self.f(states, dones, next_states) - log_pis

    def calculate_reward(self, states, dones, log_pis, next_states):
        with torch.no_grad():
            logits = self.forward(states, dones, log_pis, next_states)
            return -F.logsigmoid(-logits)
`

In the AIRL paper the algorithm states, that the reward function should be Log(D) - Log((1 - D), when training an arbitrary RL agent. It also seems like Log(D) - Log((1 - D) generates stronger gradients at the start of training compared - Log((1 - D). I tried reading the paper on AIRL, along with the GAN paper by Ian J. Goodfellow et al. But could not find any grounds for using -Log(1 - D) as the reward function. Can you tell me the advantages of training with -Log(1 - D).

**Comment by sangwoo0123** [comment]
Does anyone know the answer to this? I've been wondering the same thing. The paper formulation states log D -log (1-D).

**Comment by achimvp** [comment]
The reward function -log(1-D) is the one proposed by GAIL and per se not valid for AIRL. However one can still use it for AIRL as is done by others (https://github.com/Garyzyr001/rethinking-airl/). The 'advantage' of -log(1-D) is that it is always positive and therefore for survival environments like Ant, HalfCheetah, Walker2d, etc. we have an inductive bias. However for environments which terminate on success like MountainCarContinuous this learns a suboptimal policy (DISCRIMINATOR-ACTOR-CRITIC:  ADDRESSING SAMPLE INEFFICIENCY AND REWARD  BIAS IN ADVERSARIAL IMITATION LEARNING). Also if you have a look at the code for the AIRL paper, you realize that even there they do not use f - log(pi) but just f with a little bit of entropy regularization. So even in the original paper the authors do not stick to the reward function they suggest in the paper. 
So I think it is okay to use -log(1-D) but be aware of the bias for environments which terminate on success.
